### PR TITLE
Refactor ASTJSONTest to reduce redundancy

### PR DESCRIPTION
**Pull request by Marenz**
This is also preparation work for #9364

### CODE REVIEW

File: test/libsolidity/ASTJSONTest.cpp [hunk 1]
```diff
@@ -131,93 +131,95 @@ TestCase::TestResult ASTJSONTest::run(ostream& _stream, string const& _linePrefi
...
+		"legacy",
+		_stream,
+		_linePrefix,
+		_formatted
+	);
+
+	return resultsMatch ? TestResult::Success : TestResult::Failure;
+}
+
+bool ASTJSONTest::runTest(std::string& _expectation, std::string& _result, std::map<std::string, unsigned> const& _sourceIndices, CompilerStack& _compiler, bool _legacy, std::string const& _variation, std::ostream& _stream, std::string const& _linePrefix, bool const _formatted)
```
Comment (mijovic) [inline test/libsolidity/ASTJSONTest.cpp]: Please remove `std::` here and split this line to multi lines.

**Comment by mijovic** [comment]
I’ve rerun timed out tests

**Comment by mijovic** [comment]
There is a commit which doesn't belong here, I think you shoud rebase once more

**Comment by Marenz** [comment]
Not sure how and why that happened, but it's gone now

File: test/libsolidity/ASTJSONTest.h [hunk 2]
```diff
@@ -19,6 +19,7 @@
 #pragma once
 
 #include <libsolutil/AnsiColorized.h>
+#include <libsolidity/interface/CompilerStack.h>
```
Comment (chriseth) [inline test/libsolidity/ASTJSONTest.h]: Can you make this a forward declaration? CompilerStack.h pulls in many headers.

**Comment by chriseth** [comment]
This is quite difficult to review. Can you summarize what you did?

File: test/libsolidity/ASTJSONTest.h [hunk 3]
```diff
@@ -41,6 +42,8 @@ class ASTJSONTest: public TestCase
 	void printSource(std::ostream& _stream, std::string const& _linePrefix = "", bool const _formatted = false) const override;
 	void printUpdatedExpectations(std::ostream& _stream, std::string const& _linePrefix) const override;
 private:
+	bool runTest(std::string& _expectation, std::string& _result, std::map<std::string, unsigned> const& _sourceIndicies, CompilerStack& _compiler, bool _legacy, std::string const& _variation, std::ostream& _stream, std::string const& _linePrefix = "", bool const _formatted = false);
```
Comment (mijovic) [inline test/libsolidity/ASTJSONTest.h]: I think this line is also way too big, can you make it multi line as well?

**Comment by Marenz** [comment]
> This is quite difficult to review. Can you summarize what you did?

Sure. Basically we have twice the same code with only minor variations in run() (once for normal, once for legcay AST). I moved that into a function that we simply call twice.

For stopAfter we use that functionality to call it a third time for "stopAfter=parsing"

File: test/libsolidity/ASTJSONTest.cpp [hunk 4]
```diff
@@ -131,93 +131,106 @@ TestCase::TestResult ASTJSONTest::run(ostream& _stream, string const& _linePrefi
...
+		_linePrefix,
+		_formatted
+	) && resultsMatch;
+
+	return resultsMatch ? TestResult::Success : TestResult::Failure;
+}
+
+bool ASTJSONTest::runTest(
+	string& _expectation,
+	string& _result,
```
Comment (mijovic) [inline test/libsolidity/ASTJSONTest.cpp:174]: Do we need these two parameters? I see that you pass private members. You can distinguish which set of parameters to use by checking `_legacy` bool param.

File: test/libsolidity/ASTJSONTest.cpp [hunk 5]
```diff
@@ -131,93 +131,106 @@ TestCase::TestResult ASTJSONTest::run(ostream& _stream, string const& _linePrefi
...
+	map<string, unsigned> const& _sourceIndices,
+	CompilerStack& _compiler,
+	bool _legacy,
+	string const& _variation,
```
Comment (mijovic) [inline test/libsolidity/ASTJSONTest.cpp:178]: Maybe this parameter can be calculated based on other parameters, so it can be removed?

File: test/libsolidity/ASTJSONTest.cpp (hunk 4 above)
Comment (Marenz) [reply test/libsolidity/ASTJSONTest.cpp:174]: Sure, this might work at this point, but it reduces the modularity a lot, if you have a look at the PR that builds on that you see what I mean: 

https://github.com/ethereum/solidity/pull/9364/commits/13638f46a26f17df6ffa564d2c154d3fca11769a#diff-4296f70f815b6207bf7ae36fa1ef6a2bqfqf0-

File: test/libsolidity/ASTJSONTest.cpp (hunk 5 above)
Comment (Marenz) [reply test/libsolidity/ASTJSONTest.cpp:178]: See my other response

**Review by chriseth — APPROVED** [review]
